### ISSUE TITLE
New API version with less magic

### ISSUE THREAD

**Issue by aeplay** [garden-co/classic-jazz#714]
[current sketch](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBA0gSzAExgXhgZQJ4FsBGIANgHQBmIATgBQBEAVgIYBezAXANZLK0CUA3ACgo2AA4BTeN3QwREkGSkohg4EUYQIMAIL5olRsCiZgAC3G5GMAN6CY9mAG1EKALpsY+pAHMZtRmDYtIJ2DtCMUAjAMMjiEMCUCKJQVAA8mOJEiuIAHlDiKFrWYOIA7tS8HgDCIACyjKIAvgB81FCmCBAeGVmVMACSkFABwOIAKmLi6ZlkzTahDg6U4lAArpRgMCWlsh0QFUKLjYLHquqaWFCJYN4m5pYwufmFOnpXhsZmFlbWTi7Irj8XhutBgGhgoCGMGOag0WgAcqsCOJKHdvo88gVkFpdPoPmiHr9nNxARhaGAkfgUaDwZDoNCQnTYIoMLZFsDvN0rj4CYwADQLLaUlEeRHI1FfSwnRngekoGRkEgU8UkWLxRLJKiM85aGr1UQY57Y154oy8+aLYluPygSyiGlaJkhdnDSLRIggRjIaZZQ1YorbCrVOoNFptPbdGZ8mAIZAeDl9AAynti3sG4TAowmEh9szmbMWDnalBAO22MAAopQSzRySBYNgVjHcKIiBYCvkeLxBcdTrCLjVE51YE9-Sb3mbJVYLQ4rQCbcRhw6IbKoM6wq6ojAAPJkVL9P0vYplINYKdhhD5XAefp9Y-lPqD4e7-dDEbjSb75rNRozwswZY1g2CEdUPY0n3pAt-0WS8LBkWDcEFI4e2lYRJhgCCoBfA9RxeXk5gwTCYAAMhsBCPHvU9+kaU4zjhDCQAAMXEcRUFw41cQnT57mnKC-hJBcyBYngwUdVdULkSQAAVElwS8EAAN3Ec0MEkhRLmuW4pxgAAfWRJnUsUqQlHiVEkhj9RUmx+LcDxaFtBpQWOczMKsol-ncGB7MXaBaGjciYBkhA5MiJTzT0yjHxAAA1RgiFWZTz2c9CamY1i3JsgE7NAITWL8mMrw8IKQsUxKeN0mxAyi2L4rK74aJCNCJAYmqEqsvUGnChih2gLrUuE3kQnM9qYritrtL04r5LCqczPQgBVMAPS9VjUjGAibAAEnCfI7NWJaUzy6MAH1jskjwxmjTamFYCjY3jbkbl-U5zKm0LxBfTAwK0N7SvwmR02GTMPxzTBmjnQF2K0WgOWCaCYAAfk8R7vCQhwbzfYHsymMGIe+rzlWMuH4YcJHCZRNH-w8EolMoIb0OTFbvTGfGalaurLA26hBRZqGLM6qdKcRv8SeccRsBjTYOHF9SWdIjlPLGMXsEhzEXl+maeKF+GkY1j69yVuBxdcZptegi7ldVo0AxPaqxo5xhjhJ53hcZ1NXwzLNP0N43vwqxblvdwH32xtbLe-M2XfRrZxFpoWnfNmBebV8CQB67jvkj4WlYAcgQnOrbHPXBqjxZddk6b9bDvOrwL8HXCzxYPGoN3VuDrHvccGuLDruY9IDw600xr2c1z-OTeaXhHAbqOLtZpiBsF0vSaTrvx-x4ul+XwskesRxDFAfaoAezTPL1l8x9rk2E+36O94PkAj5PnxPJbwePaBkepkvnuJ-9g6mYfxDp3buuBe68BvqXamscKa8BIiLbarpxB2UDkdQUm0uAoAth5dBN12DWHusjTS0Z6zmEoHZSw3goj5VWKIZAERkEwAqOgfMDVTgAHp2EwHmpgbQABxCs2p6KYFdKsLQfMOoGj4vkPICoSAckFIwbwkgMCKnJnTF66FMAswwK3b0IiIhiNNnRC4MkFAIDbPPfUIsdpiJkAYtYEAJLoSkjomAejUhmLIBY8QpsmQwFECWbxliMBeJ8SQQOdBQDHQgBwegyAyAdAAOwABYADMfBGomK0OnXcYS2ziJTrqNOw4RYIRkPk8QzjmqJjcR43JZBKkQGMf2HJz5GlBJ8RAAATPPdOJBdzUEqXA6wmialjF6bo9+DSmndOMdkjCCBKBqEkBIkMUjBSBPMQUmQmEBlkCGZ0ts3YxmSETFUGQHiqhLJWabcy5zEwyArDkNQqxYipHOV3LZwS4gF2jAPQBARsARyERcNKbEikMXBWUq89jRFONOe4xilz37gvma0mA+pRA+HngAJXEKASgqA+LSwlqo+RKNBQKXtjIcFJwgA)

but:

* only one syntax for CoLists, more regular:

```
class Chat extends CoList {
  item = Message
}
```

**Comment by Schniz** [comment]
what about avoiding the "instance" syntax in favor of partial application (effect.ts-style)? this way we can have compile time validations for field names, and avoid generic static methods when they're not really generic but bound to the declaration (like `.create`): https://tsplay.dev/wQQ6jw

```ts
class Message extends CoMap.Of<Message>()({
  text: co.string,
}) {}
```

I agree that it's not the prettiest thing (the example of `class Chat extends CoList { item = Message }` is much nicer) but it's practical and very powerful

**Comment by gdorsi** [comment]
> what about avoiding the "instance" syntax in favor of partial application

Which use cases are you imagining for the partial application in lists?

**Comment by gdorsi** [comment]
## Explainer 

The major difference with the current APIs is how the CoValues are consumed and represented as "values".

Now the load functions are returning CoValues as instances of the schema class, requiring us to use Proxies and do magic in order to make Typescript types to work correctly.

We are going to change this in order to have more flexibility on what to return when a user reads a CoValue.

So APIs like `useCoState` are going to never return `undefined` but the return type will be a uninon like this one:

```ts
type Unloaded<T> = {$state: "unloaded", $jazz: { id: string }}
type Loaded<T, D> = {
    $state: "loaded",
    $kind: T[Kind]
    $jazz: {id: string, other: "magic", update: UpdateFn<T>}
} & ValueOf<T> & DepthResolution<D>
```

Example:
```ts
const [message] = useCoState(Message, messageId, {})

if (message.$state === "unloaded") return null

return <MessageContent content={message.text} author={message.$jazz.createdBy} />
```

Another important difference will be on the update API.
Instead of using seamless mutations we will move toward an `update` function.

The `update` function will work with an API similar to [Immer's produce](https://immerjs.github.io/immer/produce#example):

```ts
const [item, updateItem] = useCoState(TodoItem, id, {})

function handleChange(evt) {
  updateValue(item => {
    item.value = evt.target.value
  })
}
```

Like the `Loaded` type suggests it will also be possible to call the update function via `$jazz.update` for which the callback returned by `useCoState` is just a shortcut.

This new iteration of the API is more friendly with the frontend frameworks because:
- the React Compiler was flagging our update system as a violation of the Rules of React. It wasn't because the updates were immutable but on the compiler perspective there is no way to tell the difference between a mutation and a "proxied" mutation like ours.
- returning proxied values was making harder for us to integrate with the rest of the main frameworks (Vue, Svelte and Solid). Switching plain objects should make it way easier.

**Comment by gdorsi** [comment]
@aeplay 

We might add an API that triggers suspense and always return the loaded/fresh value:

```ts
const [loadingItem, updateItem] = useCoState(TodoItem, id, {})

const item = loadingItem.$use(); // always Loaded
```

This would make the code much more cleaner because there would be:
- No type guards to add to check if the value is unloaded
- No load state to manage internally the component
- The rejection would be handled with an ErrorBoundary

**Comment by gdorsi** [comment]
We should also have a dedicated state for the CoValues that can't be accessed for a lack of permissions.

e.g. by returning `$state = 'access-not-allowed'`

**Comment by gdorsi** [comment]
Another important point is that we are going to remove the auto-subscription that currently works by accessing to CoValues refs that aren't declared in the deep loading object.

We found that it creates confusion and relies on Proxy which makes our integration with non-react frameworks less efficient.

But, as suggested by one of our main adopters, we should still keep the ability of downloading part of the data in background and not making it render-blocking.

This is the greatest perk of the auto-subscription, but since it's too magical most of the time devs don't realize that they are using it. 

We can keep this feature by adding an API that achieves the same result but in a more explicit manner:

Example
```ts
const [issue] = useCoState(Issue, id, {})

return <>
<h1>{issue.title}</h1> {/* This is rendered before we start loading the comments */}
{issue.comments.map((comment) => {
  // This subscribes to the comment content and returns `Unloaded` until everything declared in the resolve is fully loaded
  const data = comment.$subscribe({ author: { profile: {} } }); 
```

This way the data flow will become more clear, Typescript more friendly, and the refactoring in multiple components can be easier.

**Comment by gdorsi** [comment]
Superseded by https://github.com/garden-co/jazz/pull/1547
